homeassistant/components/varta_storage/__init__.py

```diff
@@ -0,0 +1,252 @@
+"""The VARTA Storage integration."""
+from __future__ import annotations
+
+import asyncio
+from datetime import timedelta
+
+import async_timeout
+from vartastorage import vartastorage
+
+from homeassistant import config_entries, core
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    ENERGY_KILO_WATT_HOUR,
+    ENERGY_WATT_HOUR,
+    PERCENTAGE,
+    POWER_WATT,
+    TIME_HOURS,
+    Platform,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import _LOGGER, DOMAIN
+
+PLATFORMS: list[Platform] = [Platform.SENSOR]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up VARTA Storage from a config entry."""
+    varta = vartastorage.VartaStorage(entry.data["host"], entry.data["port"])
+    try:
+        varta.client.connect()
+    except Exception as ex:
+        _LOGGER.warning("Could not connect to modbus server")
+        raise ConfigEntryNotReady from ex
+
+    hass.data.setdefault(DOMAIN, {})
+    hass_data = dict(entry.data)
+
+    # Registers update listener to update config entry when options are updated.
+    unsub_options_update_listener = entry.add_update_listener(options_update_listener)
+
+    # Store a reference to the unsubscribe function to cleanup if an entry is unloaded.
+    hass_data["unsub_options_update_listener"] = unsub_options_update_listener
+
+    async def async_update_data():
+        """Fetch data and preo-process the data from API endpoint."""
+
+        def sync_update():
+            """Utilizing synchronous task as the used PyPI Package is not built with async."""
+            varta = vartastorage.VartaStorage(entry.data["host"], entry.data["port"])
+            # Collect all data from the device at once
+            varta.get_all_data()
+            return varta
+
+        try:
+            async with async_timeout.timeout(10):
+
+                # Call synchronous task to update the sensor values
+                result = await hass.async_add_executor_job(sync_update)
+        except Exception as failed_update:
+            raise UpdateFailed("Error communicating with API") from failed_update
+
+        thisdict = {}
+        thisdict[0] = {
+            "name": "VARTA state of charge",
+            "native_value": result.soc,
+            "device_class": "battery",
+            "state_class": "measurement",
+            "native_unit_of_measurement": PERCENTAGE,
+        }
```

Comment:
This is entity metadata, for which we have `EntityDescription`s in Home Assistant (search the codebase, there are many examples of these around).

There is no need to constantly update/add all this data in the coordinator, as it is a fairly static set of data.

Reply:
EntityDescription